### PR TITLE
Feat/utae

### PR DESCRIPTION
**Pull request by xiaoshenlong17**
# 卫星时间序列影像分割：U-TAE

## 案例简介
本案例实现了 **U-TAE (Unsupervised Temporal Attention Encoder)** 模型在 **PASTIS 数据集** 上的语义分割与全景分割任务。  
该模型最早发表于 *ICCV 2021*（Segmentation of Satellite Image Time Series with Convolutional Temporal Attention Networks）。  
我们将原始 PyTorch 实现迁移至 **PaddlePaddle**，并提供完整的训练与测试脚本。  

## 数据集
- **PASTIS**：一个专为遥感时间序列影像语义分割设计的数据集，包含 Sentinel-2 多时相影像及对应的地块标注。  
- 任务类型：
  - **语义分割 (Semantic Segmentation)**
  - **全景分割 (Panoptic Segmentation)**

数据集可通过 [PASTIS 官网](https://zenodo.org/records/5012942) 下载。  

## 模型结构
- **U-TAE Backbone**：采用卷积 + 时间注意力机制对卫星影像时序特征进行建模。  
- **ConvLSTM & LTAE**：用于时序特征编码与解码。  
- **Panoptic Head**：在语义分割结果上进一步进行全景分割。  

迁移实现包括以下核心组件：
- `src/backbones/`：网络骨干（ConvLSTM, LTAE, Positional Encoding, UTAE）  
- `src/learning/`：训练过程相关（loss、miou、权重初始化）  
- `src/panoptic/`：全景分割模块（PaPs, loss, FocalLoss, metrics, utils）  
- 训练/测试脚本：`train_semantic.py`、`test_semantic.py`、`train_panoptic.py`、`test_panoptic.py`

### 语义分割任务
训练：
```bash
python train_semantic.py --config configs/utae_semantic.yaml
```

测试：
```bash
python test_semantic.py --config configs/utae_semantic.yaml --weights output/utae_semantic/best_model.pdparams
```

### 全景分割任务
训练：
```bash
python train_panoptic.py --config configs/utae_panoptic.yaml
```

测试：
```bash
python test_panoptic.py --config configs/utae_panoptic.yaml --weights output/utae_panoptic/best_model.pdparams
```

## 实验结果
在 PASTIS 数据集上，本案例复现了以下性能（PaddlePaddle 实现）：  

- **SQ (Segmentation Quality)**: 83.8  
- **RQ (Recognition Quality)**: 58.9  
- **PQ (Panoptic Quality)**: 49.7  

对比 PyTorch 原版实现，性能相近，说明迁移有效。

## 文件结构
```
UTAE
 ├── src
 │   ├── backbones/              # 模型骨干
 │   ├── learning/               # 训练相关工具
 │   ├── panoptic/               # 全景分割模块
 │   ├── dataset.py              # 数据加载
 │   ├── model_utils.py          # 模型工具函数
 │   └── utils.py                # 通用工具
 ├── train_semantic.py           # 语义分割训练
 ├── test_semantic.py            # 语义分割测试
 ├── train_panoptic.py           # 全景分割训练
 └── test_panoptic.py            # 全景分割测试
```

### CODE REVIEW

File: examples/UTAE/src/backbones/ltae.py [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+            n_head (int): Number of attention heads.
+            d_k (int): Dimension of the key and query vectors.
+            mlp (List[int]): Widths of the layers of the MLP that processes the concatenated outputs of the attention heads.
+            dropout (float): dropout
+            d_model (int, optional): If specified, the input tensors will first processed by a fully connected layer
+                to project them into a feature space of dimension d_model.
+            T (int): Period to use for the positional encoding.
+            return_att (bool): If true, the module returns the attention masks along with the embeddings (default False)
+            positional_encoding (bool): If False, no positional encoding is used (default True).
+        """
```
Comment (HydrogenSulfate) [inline examples/UTAE/src/backbones/ltae.py]: class的docstring请移动到类的下方，`__init__`的上方

File: docs/zh/examples/UTAE.md [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+## 数据集介绍
+PASTIS数据集，该数据集由2433个10×128×128形状的多光谱图像序列组成。每个序列包含2018年9月至2019年11月之间的38至61个观察点，总计超过20亿像素。获取间隔时间不均匀，平均为5天。这种缺乏规律性的现象是由于卫星数据提供商对大量云层覆盖的采集进行了自动过滤。该数据集覆盖4000多平方公里，图像来自法国四个不同地区，气候和作物分布多样。
+数据集可通过 [PASTIS官网](https://zenodo.org/records/5012942) 下载。
+
+## 模型构建
+1 模型构建
+本案例基于 UTAE（U-TAE） 实现，用 PaddleScience 封装如下：
+```
+--8<--
+examples/utae/src/model.py:1:80
```
Comment (HydrogenSulfate) [inline docs/zh/examples/UTAE.md]: 类似的路径跟本PR上传的路径对不上，所以文档里都是无法显示的，麻烦根据PR的文档预览链接，修改一下，

Image

File: docs/zh/examples/UTAE.md [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+wget -O pretrained/utae_panoptic.pdparams \
+  https://paddle-org.bj.bcebos.com/paddlescience/models/utae/panoptic.pdparams
+python test_panoptic.py \
+  --weight_file ./pretrained/utae_panoptic.pdparams \
+  --dataset_folder "/path/to/PASTIS" \
+  --batch_size 2 \
+  --num_workers 0 \
+  --device gpu
+```
+
```
Comment (HydrogenSulfate) [inline docs/zh/examples/UTAE.md]: 这块的代码建议放到开头位置，格式参考：https://paddlescience-docs.readthedocs.io/zh-cn/latest/zh/examples/shock_wave/?h=%E6%BF%80%E6%B3%A2#1

File: docs/zh/examples/UTAE.md [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+## 实验结果
+在 PASTIS 数据集上，本案例复现了以下性能（PaddlePaddle 实现）：
+- **SQ (Segmentation Quality)**: 83.8  
+- **RQ (Recognition Quality)**: 58.9  
+- **PQ (Panoptic Quality)**: 49.7  
+![rusult](https://paddle-org.bj.bcebos.com/paddlescience/docs/utae/rusult.png)
```
Comment (HydrogenSulfate) [inline docs/zh/examples/UTAE.md]: 同理，实验结果和预训练权重，建议放到文档开头位置，参考：https://paddlescience-docs.readthedocs.io/zh-cn/latest/zh/examples/shock_wave/?h=%E6%BF%80%E6%B3%A2#1

File: examples/UTAE/src/backbones/convlstm.py [hunk 5]
```diff
@@ -0,0 +1,262 @@
...
+            Height and width of input tensor as (height, width).
+        input_dim: int
+            Number of channels of input tensor.
+        hidden_dim: int
+            Number of channels of hidden state.
+        kernel_size: (int, int)
+            Size of the convolutional kernel.
+        bias: bool
+            Whether or not to add the bias.
+        """
```
Comment (HydrogenSulfate) [inline examples/UTAE/src/backbones/convlstm.py]: `__init__`相关代码请移动到类名的下方

**Review by HydrogenSulfate — CHANGES_REQUESTED** [review]
有几处问题还麻烦修改下

File: docs/zh/examples/UTAE.md [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+examples/UTAE/src/backbones/utae.py:12:177
+--8<--
+```
+
+## 可视化结果
+在 PASTIS 数据集上，本案例复现了以下性能（PaddlePaddle 实现）：
+
+- **SQ (Segmentation Quality)**: 83.8  
+- **RQ (Recognition Quality)**: 58.9  
+- **PQ (Panoptic Quality)**: 49.7
```
Comment (liaoxin2) [inline docs/zh/examples/UTAE.md]: 在PASTIS 数据集上，本案例复现了全景分割预测与语义分割预测的可视化结果如图所示：

File: docs/zh/examples/UTAE.md [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+
+上图展示了 PASTIS 数据集上的农田地块分割结果。在图中用不同颜色表示不同的地块。绿色圈出的位置代表大块地被错误识别为单一地块；红色圈出的位置代表很多细长地块未被正确检测；蓝色圈出的位置展示了 全景分割优于语义分割的情况。模型在区域边界检测方面具有较好表现，尤其在复杂边界的恢复上有所优势。但在面对细长、破碎或复杂地块时，仍然存在挑战，容易导致置信度下降或检测失败。
+## 参考文献
+
+- U-TAE 原论文：[Panoptic Segmentation of Satellite Image Time Series with Convolutional Temporal Attention Networks](https://arxiv.org/abs/2107.07933)
+- 源代码实现：[python test_semantic.py \
+  --weight_file /home/aistudio/1/results/Fold_1/model_epoch_8_miou_0.477.pdparams \
+  --dataset_folder "/home/aistudio/PASTIS" \
+  --device gpu
+  --num_workers 0](https://github.com/VSainteuf/utae-paps)
```
Comment (liaoxin2) [inline docs/zh/examples/UTAE.md]: [ ]中的内容换成https://github.com/VSainteuf/utae-paps

**Review by HydrogenSulfate — APPROVED** [review]
LGTM
